### PR TITLE
[CU-2vtgc9d] Fix replay validation service failure if player mapping fails.

### PR DESCRIPTION
**Pull request by mwbrown**

### CODE REVIEW

**Comment by 0c370t** [comment]
Task linked: [CU-2vtgc9d Submitting replays that contain users not in the database breaks](https://app.clickup.com/t/2vtgc9d)

File: microservices/submission-service/src/replay-validation/replay-validation.service.ts [hunk 1]
```diff
@@ -123,18 +123,21 @@ export class ReplayValidationService {
...
             accountId: p.discordId,
         })));
-        if (userIdsResponses.some(r => r.status === ResponseStatus.ERROR)) {
+
+        // Add a check if any of the promises were rejected. This makes it safe to map them later as all PromiseFulfilledResult<T>.
+        if (userIdsResponses.some(r => r.status === "rejected" || r.value.status == ResponseStatus.ERROR)) {
             this.logger.error(`Unable to validate submission, couldn't map from MLE player to Sprocket user by discordId`, JSON.stringify(userIdsResponses));
             return {
                 valid: false,
                 errors: [],
```
Comment (bwpikaard) [inline microservices/submission-service/src/replay-validation/replay-validation.service.ts:137]: Include the full list of users that failed to be found in the error array. users.filter(same as .some statement).map(user => discordId)

**Review by bwpikaard — CHANGES_REQUESTED** [review]

File: core/src/mledb/mledb-player/mledb-player.controller.ts [hunk 2]
```diff
@@ -52,6 +52,21 @@ export class MledbPlayerController {
...
     async getPlayersByPlatformIds(@Payload() payload: unknown): Promise<CoreOutput<CoreEndpoint.GetPlayersByPlatformIds>> {
         const platformIds = CoreSchemas.GetPlayersByPlatformIds.input.parse(payload);
-        return Promise.all(platformIds.map(async d => this.getPlayerByPlatformId(d)));
+        const platformIdsResponse = await Promise.allSettled(platformIds.map(async d => this.getPlayerByPlatformId(d)));
+
+        if (platformIdsResponse.every(r => r.status === "fulfilled")) {
+            return platformIdsResponse.map(r => (r as PromiseFulfilledResult<CoreOutput<CoreEndpoint.GetPlayerByPlatformId>>).value);
+        } else {
+            // Build up a custom error response containing the list of failed platforms + IDs.
+            var failedPlatforms: string[] = [];
```
Comment (bwpikaard) [inline core/src/mledb/mledb-player/mledb-player.controller.ts:61]: var bad, no var 😞 
```suggestion
            const failedPlatforms: string[] = [];
```

File: microservices/submission-service/src/replay-validation/replay-validation.service.ts [hunk 3]
```diff
@@ -127,11 +127,24 @@ export class ReplayValidationService {
             accountType: "DISCORD",
             accountId: p.discordId,
         })));
-        if (userIdsResponses.some(r => r.status === ResponseStatus.ERROR)) {
-            this.logger.error(`Unable to validate submission, couldn't map from MLE player to Sprocket user by discordId`, JSON.stringify(userIdsResponses));
+
+        if (userIdsResponses.some(r => r.status == ResponseStatus.ERROR)) {
+
+            // Build up a list of failed player Discord IDs to return in the error response.
+            var errors: ValidationError[] = [];
```
Comment (bwpikaard) [inline microservices/submission-service/src/replay-validation/replay-validation.service.ts:134]: `var`y bad!
```suggestion
            const errors: ValidationError[] = [];
```

File: microservices/submission-service/src/replay-validation/replay-validation.service.ts [hunk 4]
```diff
@@ -127,11 +127,24 @@ export class ReplayValidationService {
...
+            userIdsResponses.forEach((r,i) => {
+                if (r.status == ResponseStatus.ERROR) {
+                    errors.push({
+                        error: `Could not find a Sprocket account for ${players[i].discordId}`,
+                        playerIndex: i,
+                    });
+                }
+            });
+
+            this.logger.error(`Unable to validate submission, couldn't map from MLE player(s) to Sprocket user by discordId:`, JSON.stringify(userIdsResponses));
```
Comment (bwpikaard) [inline microservices/submission-service/src/replay-validation/replay-validation.service.ts:144]: I wouldn't stringify here. Mapping and joining into a string is cleaner/easier to read.

**Review by bwpikaard — CHANGES_REQUESTED** [review]

File: microservices/submission-service/src/replay-validation/replay-validation.service.ts [hunk 5]
```diff
@@ -127,11 +127,24 @@ export class ReplayValidationService {
             accountType: "DISCORD",
             accountId: p.discordId,
         })));
-        if (userIdsResponses.some(r => r.status === ResponseStatus.ERROR)) {
-            this.logger.error(`Unable to validate submission, couldn't map from MLE player to Sprocket user by discordId`, JSON.stringify(userIdsResponses));
+
+        if (userIdsResponses.some(r => r.status == ResponseStatus.ERROR)) {
+
+            // Build up a list of failed player Discord IDs to return in the error response.
+            const errors: ValidationError[] = [];
```
Comment (bwpikaard) [inline microservices/submission-service/src/replay-validation/replay-validation.service.ts:134]:
```suggestion
            // Build up a list of failed player Discord IDs to return in the error response.
            const sprocketUserErrors: ValidationError[] = [];
            
```

File: microservices/submission-service/src/replay-validation/replay-validation.service.ts [hunk 6]
```diff
@@ -127,11 +127,24 @@ export class ReplayValidationService {
...
+            userIdsResponses.forEach((r,i) => {
+                if (r.status == ResponseStatus.ERROR) {
+                    errors.push({
+                        error: `Could not find a Sprocket account for ${players[i].discordId}`,
+                        playerIndex: i,
+                    });
+                }
+            });
+
+            this.logger.error(`Unable to validate submission, couldn't map from MLE player(s) to Sprocket user by discordId:`, JSON.stringify(userIdsResponses));
```
Comment (bwpikaard) [inline microservices/submission-service/src/replay-validation/replay-validation.service.ts:144]:
```suggestion
            sprocketUserErrors.forEach(err => this.logger.error(err));
```

File: microservices/submission-service/src/replay-validation/replay-validation.service.ts [hunk 7]
```diff
@@ -127,11 +127,24 @@ export class ReplayValidationService {
...
-        if (userIdsResponses.some(r => r.status === ResponseStatus.ERROR)) {
-            this.logger.error(`Unable to validate submission, couldn't map from MLE player to Sprocket user by discordId`, JSON.stringify(userIdsResponses));
+
+        if (userIdsResponses.some(r => r.status == ResponseStatus.ERROR)) {
+
+            // Build up a list of failed player Discord IDs to return in the error response.
+            const errors: ValidationError[] = [];
+            userIdsResponses.forEach((r,i) => {
+                if (r.status == ResponseStatus.ERROR) {
+                    errors.push({
```
Comment (bwpikaard) [inline microservices/submission-service/src/replay-validation/replay-validation.service.ts:137]:
```suggestion
                    sprocketUserErrors.push({
```

File: microservices/submission-service/src/replay-validation/replay-validation.service.ts [hunk 8]
```diff
@@ -127,11 +127,24 @@ export class ReplayValidationService {
...
+                        playerIndex: i,
+                    });
+                }
+            });
+
+            this.logger.error(`Unable to validate submission, couldn't map from MLE player(s) to Sprocket user by discordId:`, JSON.stringify(userIdsResponses));
             return {
                 valid: false,
-                errors: [],
+                errors: errors,
```
Comment (bwpikaard) [inline microservices/submission-service/src/replay-validation/replay-validation.service.ts:147]:
```suggestion
                errors: sprocketUserErrors,
```

**Review by bwpikaard — CHANGES_REQUESTED** [review]

**Review by bwpikaard — APPROVED** [review]

**Pull request merged**